packages/teleterm/src/services/config/configStore.ts

```diff
@@ -0,0 +1,65 @@
+import { z, ZodIssue } from 'zod';
+
+import { FileStorage } from 'teleterm/services/fileStorage';
+import Logger from 'teleterm/logger';
+
+const logger = new Logger('ConfigStore');
+
+export function createConfigStore<
+  Schema extends z.AnyZodObject,
+  Shape = z.infer<Schema>
+>(schema: Schema, defaults: Shape, fileStorage: FileStorage) {
+  const parsed = parse(fileStorage.get());
+  updateConfigFile();
+
+  function get<K extends keyof Shape>(
+    key: K
+  ): { value: Shape[K]; metadata: { isStored: boolean } } {
+    const stored = fileStorage.get<Shape>();
+    return {
+      value: { ...defaults, ...stored }[key],
+      metadata: { isStored: stored[key] !== undefined },
+    };
+  }
+
+  function set<K extends keyof Shape>(key: K, value: Shape[K]): void {
+    fileStorage.putKey(key as string, value);
+  }
+
+  function getParsingErrors(): ZodIssue[] | undefined {
```

Comment:
`getParsingErrors` can be a bit confusing I would like to just use a field here, but I can't because of IPC.
Maybe renaming it to `readValidationErrors` makes it more clear? 

I also removed updating the file.

Reply:
> `getParsingErrors` can be a bit confusing I would like to just use a field here, but I can't because of IPC. Maybe renaming it to `readValidationErrors` makes it more clear?

Ah, I forgot about IPC. As you said, Without IPC even just `validationErrors` would look fine but it looks weird with IPC.

---

What about `getStoredConfigErrors`? This would make it even more clear that `getStoredConfigErrors` returns errors on which `get`/`set` have no bearing.

For me the biggest issue about this piece of code remains to be the fact that we do some side effects in one function call but we inspect the errors from it in a totally different place instead of returning them from the function call that performs those side effects. But I realize that with IPC there's no way around it.

That's why it's important for me to make the name of it as clear as possible to bring back some of this causality which would be present if this was all done in a single process.

Reply:
`getStoredConfigErrors` sounds totally fine to me, I don't have any strong opinion on the name, so I am happy to change it.